### PR TITLE
cleanup(GCS+gRPC): protos in `Connection::*Upload()`

### PR DESCRIPTION
**Pull request by coryan**
Start using protos in the implementation of the `Connection*` classes.

Part of the work for #13910

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14008)

### CODE REVIEW

File: google/cloud/storage/internal/async/connection_impl.cc [hunk 1]
```diff
@@ -66,6 +67,25 @@ idempotency_policy(Options const& options) {
...
+      options.get<storage_experimental::EnableMD5ValidationOption>();
+
+  if (!crc32c && !md5) return storage::internal::CreateNullHashFunction();
+  if (crc32c && md5) {
+    return std::make_unique<storage::internal::CompositeFunction>(
+        std::make_unique<storage::internal::Crc32cHashFunction>(),
+        storage::internal::MD5HashFunction::Create());
+  }
+  if (crc32c) std::make_unique<storage::internal::Crc32cHashFunction>();
+  return storage::internal::MD5HashFunction::Create();
```
Comment (devbww) [inline google/cloud/storage/internal/async/connection_impl.cc:86]: Opt nit:  I'd make `return storage::internal::CreateNullHashFunction()` the default return by guarding this with `if (md5)` and deleting line 79.

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/internal/async/connection_impl.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl.cc]: Done. Also note the bug in line 85, I will be adding tests in an upcoming PR.

Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl.cc]: Do let me know if I should reorder the PRs to include this code and the tests in a single PR.

Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl.cc]: I convinced myself that I should also include the tests.  Going to pause this PR until those are ready to merge.

Comment (devbww) [reply google/cloud/storage/internal/async/connection_impl.cc]: Okie dokie.

**Review by devbww — APPROVED** [review]

**Pull request merged**